go/internal/feast/featurestore.go

```diff
@@ -0,0 +1,885 @@
+			}
+			requestedResultRowNames[strings.Replace(featureReference, ":", "__", 1)] = struct{}{}
+		}
+	} else {
+		for _, featureReference := range featureRefs {
+			parsedFeatureName := strings.Split(featureReference, ":")
+			var featureName string
+			if len(parsedFeatureName) == 0 {
+				return nil, errors.New("featureReference should be in the format: 'FeatureViewName:FeatureName'")
+			} else if len(parsedFeatureName) == 1 {
+				featureName = parsedFeatureName[0]
+			} else {
+				featureName = parsedFeatureName[1]
+			}
+			requestedResultRowNames[featureName] = struct{}{}
+		}
+	}
+
+	featureViews := make([]*FeatureView, len(groupedRefs))
+	index := 0
+	for featureView := range groupedRefs {
+		featureViews[index] = featureView
+		index += 1
+	}
+
+	// TODO (Ly): This should return empty now
+	// Expect no ODFV or Request FV passed in GetOnlineFearuresRequest
+	neededRequestData, neededRequestFvFeatures, err := fs.getNeededRequestData(groupedOdfvRefs, groupedRequestFvRefs)
+	if err != nil {
+		return nil, err
+	}
+
+	joinKeyValues := make(map[string]*types.RepeatedValue)
+	requestDataFeatures := make(map[string]*types.RepeatedValue) // TODO (Ly): Should be empty now until ODFV and Request FV are supported
+	for entityName, vals := range entityProtos {
+		if _, ok := neededRequestFvFeatures[entityName]; ok {
+			requestedResultRowNames[entityName] = struct{}{}
+			requestDataFeatures[entityName] = vals
+		} else if _, ok = neededRequestData[entityName]; ok {
+			requestDataFeatures[entityName] = vals
+		} else {
+			if joinKey, ok := entityNameToJoinKeyMap[entityName]; !ok {
+				return nil, errors.New(fmt.Sprintf("entityNotFoundException: %s\n%v", entityName, entityNameToJoinKeyMap))
+			} else {
+				joinKeyValues[joinKey] = vals
+				requestedResultRowNames[joinKey] = struct{}{}
+			}
+		}
+
+	}
+
+	// TODO (Ly): Skip this validation since we're not supporting ODFV yet
+
+	// err = fs.ensureRequestedDataExist(neededRequestData, neededRequestFvFeatures, requestDataFeatures)
+	// if err != nil {
+	// 	return nil, err
+	// }
+
+	onlineFeatureResponse := &serving.GetOnlineFeaturesResponse{Metadata: &serving.GetOnlineFeaturesResponseMetadata{
+		FeatureNames: &serving.FeatureList{Val: make([]string, 0)},
+	},
+		Results: make([]*serving.GetOnlineFeaturesResponse_FeatureVector, numRows),
+	}
+	// Allocate memory for each GetOnlineFeaturesResponse_FeatureVector
+	for index = 0; index < numRows; index++ {
+		onlineFeatureResponse.Results[index] = &serving.GetOnlineFeaturesResponse_FeatureVector{Values: make([]*types.Value, 0),
+			Statuses:        make([]serving.FieldStatus, 0),
+			EventTimestamps: make([]*timestamppb.Timestamp, 0),
+		}
+	}
+	// Merge joinKeyValues into requestDataFeatures
+	for entityName, vals := range joinKeyValues {
+		requestDataFeatures[entityName] = vals
+	}
+
+	fs.populateResultRowsFromColumnar(onlineFeatureResponse, requestDataFeatures)
```

Comment:
I think it's highly inefficient, what's being done here. I understand that it was moved here from Python, but we can fix it instead of just keeping inefficient implementation.

So, what's going on here: we copy features from request to response to make them accessible by ODFV transformations. And then at the end of this long method we delete them from response (because user actually didn't ask for these features - they are input parameters). Since these features are needed only by ODFV transformations I suggest we pass them separately  to transformations only.

Reply:
@pyalex agreed, but can we add a TODO and use a follow up PR to fix it? The scope of this PR is already massive.

Reply:
The flow in python is: Copy all provided entities + entities in ODFVs + requested FV schema into onlineFeatureResponse -> convert this to Dataframe -> use this dataframe copy and add/ remove column as needed and call udf to get transformed columns -> append to onlineFeatureResponse requested transformed values. Depending on how we want to implement ODFV, I could see a new flow as:

Create a separate dataframe for odfvs: this could either be like python's implementation where we merge all entities + entities in ODFVs + requested FV schema into 1 dataframe and remove/add to this dataframe object as input to udf (essentially function augmentResponseWithOnDemandTransforms which I have skipped for this PR) and append to online response - and we never add features not requested in online response.
We can also create a new dataframe consisting of all required inputs for each ODFV, although I'm not sure if we will end up increasing time due to copy multiple data multiple times instead of reusing and fixing 1 dataframe.

There is also a dummy entity that gets added/deleted, we can optimize on this as well and since I'll be rewriting groupRefs, I can make a note on the new flow and rewrite?